test/common/upstream/priority_conn_pool_map_impl_test.cc

```diff
@@ -0,0 +1,144 @@
+#include <memory>
+
+#include "envoy/http/conn_pool.h"
+
+#include "common/upstream/priority_conn_pool_map_impl.h"
+
+#include "test/mocks/common.h"
+#include "test/mocks/event/mocks.h"
+#include "test/mocks/http/conn_pool.h"
+#include "test/mocks/upstream/host.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using testing::AnyNumber;
+using testing::AtLeast;
+using testing::Return;
+using testing::SaveArg;
+
+namespace Envoy {
+namespace Upstream {
+namespace {
+
+class PriorityConnPoolMapImplTest : public testing::Test {
+public:
+  using TestMap = PriorityConnPoolMap<int, Http::ConnectionPool::Instance>;
+  using TestMapPtr = std::unique_ptr<TestMap>;
+
+  TestMapPtr makeTestMap() { return std::make_unique<TestMap>(dispatcher_, host_); }
+
+  TestMap::PoolFactory getBasicFactory() {
+    return [&]() {
+      auto pool = std::make_unique<NiceMock<Http::ConnectionPool::MockInstance>>();
+      ON_CALL(*pool, hasActiveConnections).WillByDefault(Return(false));
+      mock_pools_.push_back(pool.get());
+      return pool;
+    };
+  }
+
+protected:
+  NiceMock<Event::MockDispatcher> dispatcher_;
+  std::vector<NiceMock<Http::ConnectionPool::MockInstance>*> mock_pools_;
+  std::shared_ptr<NiceMock<MockHost>> host_ = std::make_shared<NiceMock<MockHost>>();
+};
+
+// Show that we return a non-null value, and that we ask invoke the default resource manager
+TEST_F(PriorityConnPoolMapImplTest, DefaultPriorityProxiedThrough) {
+  TestMapPtr test_map = makeTestMap();
+
+  EXPECT_CALL(host_->cluster_, resourceManager(ResourcePriority::High)).Times(0);
+  EXPECT_CALL(host_->cluster_, resourceManager(ResourcePriority::Default)).Times(AtLeast(1));
+
+  auto pool = test_map->getPool(ResourcePriority::Default, 0, getBasicFactory());
+  EXPECT_TRUE(pool.has_value());
+
+  // At this point, we may clean up/decrement by 0, etc, so allow any number.
+  EXPECT_CALL(host_->cluster_, resourceManager(ResourcePriority::High)).Times(AnyNumber());
+}
+
+// Show that we return a non-null value, and that we ask invoke the high resource manager
+TEST_F(PriorityConnPoolMapImplTest, HighPriorityProxiedThrough) {
+  TestMapPtr test_map = makeTestMap();
+
+  EXPECT_CALL(host_->cluster_, resourceManager(ResourcePriority::Default)).Times(0);
+  EXPECT_CALL(host_->cluster_, resourceManager(ResourcePriority::High)).Times(AtLeast(1));
+
+  auto pool = test_map->getPool(ResourcePriority::High, 0, getBasicFactory());
+  EXPECT_TRUE(pool.has_value());
+
+  // At this point, we may clean up/decrement by 0, etc, so allow any number.
+  EXPECT_CALL(host_->cluster_, resourceManager(ResourcePriority::Default)).Times(AnyNumber());
```

Comment:
How come we can't expect a specific number of calls? Is this not deterministic?

Reply:
It is, but it wasn't really relevant to the test. I prefer not to assert on things at all if they're not the main point of the test, but since this isn't a nice mock, I have to tell it something.

Why do I prefer to do that? It minimizes churn when making unrelated behavioural changes. For example, if the proxied-to class decides that increment then decrement then increment then decrement is the correct behaviour, it's not really the job of this class's UT to assert that. If I have a precise number here, every time a change in how the proxied-to class manages its resources comes in, I have to change this code.

That said I'm not dead-set on this approach. If you prefer to cast a wider testing net and assert here as well, I can change this to be more explicit.

Reply:
I'm fine with that approach. I guess I just reacted to this saying `AnyNumber`, which I think means it might not be called at all? Should it be `AtLeast(1)`? Not really a big deal

Reply:
I'll make it AtLeast(1). Seems reasonable.